### PR TITLE
Support block delete with alt and meta modifiers

### PR DESCRIPTION
**Pull request by canastro**
This PR focuses on supporting modifiers to delete blocks of text in a editable widget.

### Shortcuts per OS
* macOS
  * **CMD (⌘) + Backspace**: Delete all words left to the cursor
  * **Option + Backspace**: Delete word left to the cursor
* Non macOS
  * **Alt + Backspace**: Delete all words left to the cursor
  * **Ctrl + Backspace**: Delete word left to the cursor

### Remarkable edge cases:
* when the cursor is preceded by a break line, independently of modifiers only the last character is deleted
* when using CJK characters the option/ctrl modifier only deletes a single character   

### Related issue: 
* https://github.com/flutter/flutter/issues/57015

### Future work:
If the approach used here is approved, I'll look into the same feature but for the "forward" delete, ie. `CMD + Delete`.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -656,15 +656,34 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
-    } else if (key == LogicalKeyboardKey.delete) {
-      _handleDelete(forward: true);
-    } else if (key == LogicalKeyboardKey.backspace) {
-      _handleDelete(forward: false);
+    } 
+  }
+
+  // Return the offset at the start of the nearest word to the left of the given
+  // offset.
+  static int _getLeftByWord(TextPainter textPainter, int offset, [bool includeWhitespace = true]) {
```
Comment (canastro) [inline packages/flutter/lib/src/rendering/editable.dart:676]: I noticed that [this PR](https://github.com/flutter/flutter/pull/75032/files#diff-49a445da2fd4112ade71a9423b3acffca8d7ffcf694849159f0adb0ed40ba0d5R779) already dealt with block navigation with Editable fields, so I copied this method from there.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 2]
```diff
@@ -956,7 +975,38 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
     }
   }
 
-  void _handleDelete({ required bool forward }) {
+  /// Computes the text after a delete event with a alt or meta modifier
+  String _computeBeforeTextWithBlockBackspace({
```
Comment (canastro) [inline packages/flutter/lib/src/rendering/editable.dart]: To make it easier to read the `_handleDelete` method, I extracted `_computeBeforeTextWithBlockBackspace
` and `_computeBeforeTextWithBackspace` to different methods

File: packages/flutter/lib/src/rendering/editable.dart [hunk 3]
```diff
@@ -966,18 +1016,32 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
     // If not deleting a selection, delete the next/previous character.
     if (selection.isCollapsed) {
       if (!forward && textBefore.isNotEmpty) {
-        final int characterBoundary = previousCharacter(textBefore.length, textBefore);
-        textBefore = textBefore.substring(0, characterBoundary);
-        cursorPosition = characterBoundary;
+        final bool endsWithBreakLine = textBefore.codeUnitAt(textBefore.length - 1) == 0x0A;
+        final bool hasModifier = metaModifier || altModifier;
+        
+        if (!hasModifier || endsWithBreakLine) {
```
Comment (canastro) [inline packages/flutter/lib/src/rendering/editable.dart:1028]: when the cursor is next to a break line, then any type of delete will only delete the line break.

**Review by renancaraujo — CHANGES_REQUESTED** [review]
fodeu nosso time

File: packages/flutter/lib/src/rendering/editable.dart [hunk 4]
```diff
@@ -637,7 +637,7 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
     final Set<LogicalKeyboardKey> keysPressed = LogicalKeyboardKey.collapseSynonyms(RawKeyboard.instance.keysPressed);
     final LogicalKeyboardKey key = keyEvent.logicalKey;
 
-    final bool isMacOS = keyEvent.data is RawKeyEventDataMacOs;
+    final bool isMacOS = defaultTargetPlatform == TargetPlatform.macOS;
```
Comment (canastro) [inline packages/flutter/lib/src/rendering/editable.dart]: I wanted to be able to override platform on the unit tests, so I changed this. 
I'm not sure if this is correct, please let me know if theres a better way.

**Pull request closed without merging**